File: DSA/Gautam_Shyamnani_590013078/Day19Question1.cpp
#include <iostream>
#include <vector>
using namespace std;

int searchInRotatedArray(vector<int>& nums, int target) {
    int left = 0, right = nums.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

       
        if (nums[mid] == target)
            return mid;

       
        if (nums[left] <= nums[mid]) {
            
            if (nums[left] <= target && target < nums[mid])
                right = mid - 1;
            else
                left = mid + 1;
        } else {
          
            if (nums[mid] < target && target <= nums[right])
                left = mid + 1;
            else
                right = mid - 1;
        }
    }

    return -1; 
}

int main() {
    vector<int> nums = {4, 5, 6, 7, 0, 1, 2};
    int target = 0;
    int result = searchInRotatedArray(nums, target);

    if (result != -1)
        cout << "Target found at index " << result << endl;
    else
        cout << "Target not found" << endl;

    return 0;
}

File: DSA/Pranvkumar_590011587/Day19_Question1.py
def search_in_rotated_array(nums, target):
    # Searches for target in a rotated sorted array and returns its index, or -1 if not found.
    left, right = 0, len(nums) - 1
    while left <= right:
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        # Left half is sorted
        if nums[left] <= nums[mid]:
            if nums[left] <= target < nums[mid]:
                right = mid - 1
            else:
                left = mid + 1
        # Right half is sorted
        else:
            if nums[mid] < target <= nums[right]:
                left = mid + 1
            else:
                right = mid - 1
    return -1

# Example usage:
if __name__ == "__main__":
    # Problem 1
    print(search_in_rotated_array([4, 5, 6, 7, 0, 1, 2], 0))
    print(search_in_rotated_array([4, 5, 6, 7, 0, 1, 2], 3))  
    print(search_in_rotated_array([11, 13, 15, 17, 19, 2, 3, 6, 7], 6))  